{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "bases = ['A', 'C', 'T', 'G']\n",
    "start_codon = 'ATG'  # Fixed as a string, not a list\n",
    "stop_codons = ['TGA', 'TAG', 'TAA']\n",
    "gen_len = 51\n",
    "\n",
    "\n",
    "table = { \n",
    "            'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "            'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "            'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "            'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "            'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "            'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "            'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "            'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "            'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "            'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "            'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "            'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "            'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "            'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "            'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_', \n",
    "            'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W', \n",
    "        } \n",
    "\n",
    "hydrophilic = ['R', 'N', 'D', 'Q', 'E', 'H', 'K', 'S', 'T']\n",
    "\n",
    "hydrophobic = ['A', 'I', 'L', 'M', 'F', 'W', 'Y', 'V']\n",
    "\n",
    "neutral = ['G', 'C', 'P']\n",
    "\n",
    "stop = [\"_\"]\n",
    "\n",
    "class indiv:  # genome, fitness, constructor, print, ...\n",
    "    def __init__(self):  # constructor\n",
    "        self.fitness = 0\n",
    "\n",
    "        # generates random mutations of length 51\n",
    "        self.genome = []\n",
    "        self.amino_acid = []\n",
    "        for i in range(0, gen_len):\n",
    "            self.genome.append(random.choice(bases))\n",
    "        \n",
    "        for i in range (0, len(self.genome), 3):\n",
    "            codon = ''.join(self.genome[i:i + 3]) \n",
    "            amino_acid = table.get(codon, \"X\")\n",
    "            self.amino_acid.append(amino_acid)\n",
    "        \n",
    "        # print('Original genome = ', ''.join(self.genome))\n",
    "\n",
    "    def to_amino_acid(self):\n",
    "        self.amino_acid = []\n",
    "        for i in range (0, len(self.genome), 3):\n",
    "            codon = ''.join(self.genome[i:i + 3])  # Extract a\n",
    "            # print('codon = ', codon)\n",
    "            amino_acid_find = table.get(codon)\n",
    "            # print('amino_acid = ', amino_acid)\n",
    "            self.amino_acid.append(amino_acid_find)\n",
    "        return self.amino_acid\n",
    "            \n",
    "    def print(self):\n",
    "        print(''.join(self.genome))\n",
    "        print(\"fitness = \", self.fitness)\n",
    "\n",
    "    def __str__(self):\n",
    "        return_str = ''.join(self.genome)\n",
    "        return_str += \"\\nfitness = \" + str(self.fitness)\n",
    "        return return_str\n",
    "\n",
    "    def copy(self, source):\n",
    "        self.genome = source.genome.copy()  # Efficient copy of genome\n",
    "        self.amino_acid = source.amino_acid.copy()\n",
    "\n",
    "    # One-point crossover\n",
    "    # def crossover(self, other):\n",
    "    #     crossPoint = random.randint(0, len(self.genome) - 1)\n",
    "    #     for i in range(crossPoint, len(self.genome)):\n",
    "    #         print('temp1 = ', len(self.genome))\n",
    "    #         temp = self.genome[i]\n",
    "    #         print('temp2 = ', len(other.genome))\n",
    "\n",
    "    #         self.genome[i] = other.genome[i]\n",
    "    #         other.genome[i] = temp\n",
    "\n",
    "    def crossover(self, other):\n",
    "        # Determine the minimum length of the two genomes\n",
    "        min_len = min(len(self.genome), len(other.genome))\n",
    "        \n",
    "        # Select a random crossover point within the valid range\n",
    "        crossPoint = random.randint(0, min_len - 1)\n",
    "        \n",
    "        # Perform the crossover operation up to the minimum length\n",
    "        for i in range(crossPoint, min_len):\n",
    "            temp = self.genome[i]\n",
    "            self.genome[i] = other.genome[i]\n",
    "            other.genome[i] = temp\n",
    "\n",
    "        ### Check if the genome is divisible by 3. If not, remove the last nucleic acid\n",
    "        # because it will not be able to form a codon. Returns none for amino acid. This makes things easier\n",
    "        if len(self.genome) % 3 != 0:\n",
    "            self.genome = self.genome[:len(self.genome) - 1]\n",
    "            \n",
    "        if len(self.genome) % 3 != 0:\n",
    "            self.genome = self.genome[:len(self.genome) - 1]\n",
    "        \n",
    "        if len(self.genome) % 3 != 0:\n",
    "            self.genome = self.genome[:len(self.genome) - 1]\n",
    "        \n",
    "        self.amino_acid = self.to_amino_acid()\n",
    "        other.amino_acid = other.to_amino_acid()\n",
    "\n",
    "\n",
    "\n",
    "    def insertion(self):\n",
    "        new_gene = []\n",
    "        new_genome = []\n",
    "        start = random.randint(0, len(self.genome) - 1)\n",
    "        for i in range(3):\n",
    "            new_gene += random.choice(bases)\n",
    "        new_genome = self.genome[:start] + new_gene + self.genome[start:]\n",
    "        self.genome = new_genome\n",
    "        # print('insert_genome = ', ''.join(new_genome))\n",
    "        self.amino_acid = self.to_amino_acid()\n",
    "        # print('insert_aa = ', ''.join(self.amino_acid))\n",
    "\n",
    "\n",
    "    def deletion(self):\n",
    "        new_genome = []\n",
    "        start = random.randint(0, len(self.genome) - 1)\n",
    "        length = 3 #only 3 nucleic acids can be deleted\n",
    "        new_genome = self.genome[:start] + self.genome[start + length:]\n",
    "        self.genome = new_genome\n",
    "        # print('delete_genome = ', ''.join(new_genome))\n",
    "        self.amino_acid = self.to_amino_acid()\n",
    "        # print('delete_aa = ', ''.join(new_genome))\n",
    "\n",
    "         \n",
    "    def mutate(self):\n",
    "        first_25 = round(len(self.genome) * .25)\n",
    "        first_75 = round(len(self.genome) * .75)\n",
    "        first_100 = len(self.genome)\n",
    "\n",
    "        # print(\"first_25 = \", first_25)\n",
    "        # print(\"first_75 = \", first_75)\n",
    "        # print(\"first_100 = \", first_100)\n",
    "\n",
    "        genome_25 = self.genome[:first_25]\n",
    "        genome_75 = self.genome[first_25:first_75]\n",
    "        genome_100 = self.genome[first_75:]\n",
    "\n",
    "        for codon in range(0, first_25, 3):\n",
    "            if table.get(''.join(self.genome[codon:codon + 3])) in hydrophilic:\n",
    "                # Mutate randomly 3 genes within the first 25 percent of the genome\n",
    "                for i in range(3):\n",
    "                    self.genome[random.randint(0, first_25 - 1)] = random.choice(bases)\n",
    "            elif table.get(''.join(self.genome[codon:codon + 3])) == \"_\":\n",
    "                for i in range(3):\n",
    "                    self.genome[random.randint(0, first_25 - 1)] = random.choice(bases)\n",
    "            else:\n",
    "                pass\n",
    "            \n",
    "        for codon in range(first_25, first_75, 3):\n",
    "            if table.get(''.join(self.genome[codon:codon + 3])) in hydrophilic:\n",
    "                # Mutate randomly 3 genes within the first 25 percent of the genome\n",
    "                for i in range(3):\n",
    "                    self.genome[random.randint(first_25, first_75 - 1)] = random.choice(bases)\n",
    "            elif table.get(''.join(self.genome[codon:codon + 3])) == \"_\":\n",
    "                for i in range(3):\n",
    "                    self.genome[random.randint(first_25, first_75 - 1)] = random.choice(bases)\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "        for codon in range(first_75, first_100, 3):\n",
    "            if table.get(''.join(self.genome[codon:codon + 3])) in hydrophilic:\n",
    "                # Mutate randomly 3 genes within the first 25 percent of the genome\n",
    "                for i in range(3):\n",
    "                    self.genome[random.randint(first_75, first_100 - 1)] = random.choice(bases)\n",
    "            elif table.get(''.join(self.genome[codon:codon + 3])) == \"_\":\n",
    "                for i in range(3):\n",
    "                    self.genome[random.randint(first_75, first_100 - 1)] = random.choice(bases)\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "        self.amino_acid = self.to_amino_acid()\n",
    "\n",
    "    \n",
    "    def calcFitness(self):\n",
    "        # Fitness is the range between the start and stop codon. Fitness is counted by three.\n",
    "        self.fitness = 0\n",
    "\n",
    "        # Search for the start and stop codons\n",
    "        genome_str = self.amino_acid\n",
    "\n",
    "        for i in range(0, round(len(genome_str) *.25)):\n",
    "            if genome_str[i] in hydrophilic:\n",
    "                self.fitness += 1\n",
    "            elif genome_str[i] in hydrophobic:\n",
    "                self.fitness -= 1\n",
    "            elif genome_str[i] in stop:\n",
    "                self.fitness -= 1\n",
    "            elif genome_str[i] in neutral:\n",
    "                self.fitness += 0\n",
    "\n",
    "        for i in range(round(len(genome_str) *.25), round(len(genome_str) *.75)):\n",
    "            if genome_str[i] in hydrophilic:\n",
    "                self.fitness -= 1\n",
    "            elif genome_str[i] in hydrophobic:\n",
    "                self.fitness += 1\n",
    "            elif genome_str[i] in stop:\n",
    "                self.fitness -= 1\n",
    "            elif genome_str[i] in neutral:\n",
    "                self.fitness += 0\n",
    "\n",
    "        for i in range(round(len(genome_str) *.75), len(genome_str) - 1):\n",
    "            if genome_str[i] in hydrophilic:\n",
    "                self.fitness += 1\n",
    "            elif genome_str[i] in hydrophobic:\n",
    "                self.fitness -= 1\n",
    "            elif genome_str[i] in stop:\n",
    "                self.fitness -= 1\n",
    "            elif genome_str[i] in neutral:\n",
    "                self.fitness += 0\n",
    "\n",
    "        return self.fitness\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "popSize = 200\n",
    "class pop:\n",
    "    def __init__(self):  # constructor\n",
    "        self.population = []\n",
    "        for i in range(0, popSize):\n",
    "            self.population.append(indiv())\n",
    "        self.bestFit = 0 # best fitness\n",
    "        self.best = 0 # index of best individual\n",
    "        self.avgFit = 0\n",
    "        self.calcStats()\n",
    "\n",
    "\n",
    "    def generation(self):\n",
    "        tempPop = pop()\n",
    "        num = -1\n",
    "        for i in range(0,popSize,2):\n",
    "\n",
    "            p1 = self.tourn() # tournament selection\n",
    "            p2 = self.tourn()\n",
    "\n",
    "            tempPop.population[i].copy(self.population[p1])\n",
    "            tempPop.population[i+1].copy(self.population[p2])\n",
    "\n",
    "            num = random.randint(0,1)\n",
    "            ## Population mutation either insertion or deletion on a coin flip \n",
    "            if(num == 0):\n",
    "                tempPop.population[i].insertion()\n",
    "                tempPop.population[i+1].insertion()\n",
    "            elif(num == 1):\n",
    "                tempPop.population[i].deletion()\n",
    "                tempPop.population[i+1].deletion()\n",
    "\n",
    "            tempPop.population[i].crossover(tempPop.population[i+1])\n",
    "            \n",
    "            tempPop.population[i].mutate()\n",
    "            tempPop.population[i+1].mutate()\n",
    "        for i in range(0,popSize):\n",
    "            self.population[i].copy(tempPop.population[i])\n",
    "\n",
    "    def tourn(self):\n",
    "        best = random.randint(0,popSize-1) # the winner so far\n",
    "        bestfit = self.population[best].fitness # best fit so far\n",
    "        for i in range(5): # tournament size of 6!!!!\n",
    "            p2 = random.randint(0,popSize-1)\n",
    "            if(self.population[p2].fitness > bestfit):\n",
    "                bestfit = self.population[p2].fitness\n",
    "                best = p2\n",
    "        return best\n",
    "\n",
    "    def calcStats(self):\n",
    "        self.avgFit = 0\n",
    "        self.population[0].calcFitness()\n",
    "        self.bestFit = self.population[0].fitness\n",
    "        self.best = 0\n",
    "        for i in range(len(self.population)):\n",
    "            self.population[i].calcFitness() # update fitnesses\n",
    "            if(self.population[i].fitness > self.bestFit): # compare fitness to best\n",
    "                self.bestFit = self.population[i].fitness\n",
    "                self.best = i\n",
    "            self.avgFit += self.population[i].fitness\n",
    "        self.avgFit = self.avgFit/len(self.population)\n",
    "\n",
    "    # Function that returns a randome genome from population size when called\n",
    "    def randGenome(self):\n",
    "        return self.population[random.randint(0,popSize-1)]\n",
    "\n",
    "    def clear(self):\n",
    "        ## clear the population back to zero\n",
    "        self.population = []\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['YGLPCRPDKVYNTDTESLSAALAALCRWLG_VNSV',\n",
       " 'PQYPAPAGSNILGSVGLVSVIANFSYASTSQTPV_QQ',\n",
       " '_ANICHELSLPF_LAMPPQLVWYYD_G_RRYDEQKMI',\n",
       " 'GHNRRNRTQLRSITTDADSARYCLHNQMFSHTR',\n",
       " 'KYPRHVFGLGVVPGGLMGV_LFGAC_VVRQHSL',\n",
       " 'GESASVGPRIFSILGLVRLPAPLRYVSASTQ',\n",
       " 'RARLAPQYEELVRLAIMCMSSTAFRSLDNPW',\n",
       " 'HVD_ESKAQMAFIAHIWCFVGVCKAWEGL_INRLK',\n",
       " 'LHATKH__LLVYLTTPLIIVLLLSAGNLSSTQPIA',\n",
       " 'AKRGLYSKFVFLALAFCSGIYIRPPQSTSIQ']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gen_array = []\n",
    "with open(\"gen_array.csv\", \"r\") as genfile: \n",
    "    reader = csv.reader(genfile)\n",
    "    for row in reader:\n",
    "        gen_array.append(''.join(row))\n",
    "\n",
    "gen_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(gen_array[0])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
